File: kridarshi/farm/integrations.py
"""
Hour 3.5-5: Weather Intelligence + Market Insights
Hour 2-3.5: Real-time SMS/WhatsApp Alerts

Each function has a mock/fallback path so the demo never breaks if a
third-party API key isn't approved in time or wifi is bad on stage.
"""
import json
import os
from pathlib import Path

import requests
from django.conf import settings

DATA_DIR = Path(__file__).resolve().parent / "data"


def get_weather(lat: float, lon: float) -> dict:
    """Hour 3.5-5: OpenWeatherMap. Falls back to mock data with no API key."""
    if not settings.OPENWEATHER_API_KEY:
        return {"mock": True, "temp_c": 29, "condition": "Partly cloudy", "rain_chance_pct": 20}
    try:
        resp = requests.get(
            "https://api.openweathermap.org/data/2.5/weather",
            params={"lat": lat, "lon": lon, "appid": settings.OPENWEATHER_API_KEY, "units": "metric"},
            timeout=5,
        )
        resp.raise_for_status()
        data = resp.json()
        return {
            "mock": False,
            "temp_c": data["main"]["temp"],
            "condition": data["weather"][0]["description"],
            "rain_chance_pct": None,  # use One Call API for a real forecast/probability
        }
    except Exception:
        return {"mock": True, "temp_c": 29, "condition": "Partly cloudy (fallback)", "rain_chance_pct": 20}


import csv

def get_market_prices(crop: str = None) -> list[dict]:
    path = DATA_DIR / "market_prices.csv"
    prices = []
    with open(path, newline="", encoding="utf-8-sig") as f:  # utf-8-sig strips Excel's hidden BOM character
        reader = csv.DictReader(f)
        # normalize headers: lowercase + stripped, so "Crop", " crop", "CROP" all work
        reader.fieldnames = [h.strip().lower() for h in reader.fieldnames]

        for row in reader:
            row = {k.strip().lower(): v for k, v in row.items()}
            prices.append({
                "crop": row.get("crop") or row.get("commodity") or "Unknown",
                "market": row.get("market") or row.get("mandi") or "Unknown",
                "price_per_quintal": float(row.get("price_per_quintal") or row.get("modal_price") or row.get("price") or 0),
                "trend": row.get("trend", "—"),
                "trader_contact": row.get("trader_contact", "N/A"),
            })
    if crop:
        return [p for p in prices if p["crop"].lower() == crop.lower()]
    return prices

def send_alert(phone: str, message: str) -> bool:
    """
    Hour 2-3.5: Twilio WhatsApp/SMS. Returns False (and logs) instead of crashing
    the demo if credentials aren't set — the alert still gets created in the DB either way.
    """
    if not (settings.TWILIO_ACCOUNT_SID and settings.TWILIO_AUTH_TOKEN):
        print(f"[MOCK ALERT] to {phone}: {message}")
        return False
    try:
        from twilio.rest import Client
        client = Client(settings.TWILIO_ACCOUNT_SID, settings.TWILIO_AUTH_TOKEN)
        client.messages.create(
            body=message,
            from_=settings.TWILIO_WHATSAPP_FROM,
            to=f"whatsapp:{phone}",
        )
        return True
    except Exception as e:
        print(f"[ALERT SEND FAILED] {e}")
        return False


TTS_URL = "https://api.elevenlabs.io/v1/text-to-speech/{voice_id}"
STT_URL = "https://api.elevenlabs.io/v1/speech-to-text"
FREE_PLAN_VOICE_ID = "JBFqnCBsd6RMkjVDRZzb"


def _tts_request(api_key: str, voice_id: str, text: str, model_id: str) -> tuple[bytes | None, str, str]:
    try:
        resp = requests.post(
            TTS_URL.format(voice_id=voice_id),
            headers={
                "xi-api-key": api_key,
                "Content-Type": "application/json",
                "Accept": "audio/mpeg",
            },
            json={"text": text[:2500], "model_id": model_id},
            timeout=30,
        )
        if resp.ok:
            return resp.content, "", ""
        detail = ""
        try:
            payload = resp.json().get("detail") or {}
            if isinstance(payload, dict):
                detail = payload.get("message") or payload.get("code") or ""
            else:
                detail = str(payload)
        except Exception:
            detail = resp.text[:200]
        print(f"[ELEVENLABS TTS FAILED] {resp.status_code} {detail}")
        return None, "tts_failed", detail
    except Exception as e:
        print(f"[ELEVENLABS TTS FAILED] {e}")
        return None, "tts_failed", str(e)


def synthesize_speech(text: str) -> tuple[bytes | None, str]:
    """ElevenLabs TTS. Returns (audio_bytes, error_message)."""
    api_key = getattr(settings, "ELEVENLABS_API_KEY", "")
    if not api_key:
        return None, "missing_key"
    voice_id = settings.ELEVENLABS_VOICE_ID or FREE_PLAN_VOICE_ID
    model_id = settings.ELEVENLABS_MODEL_ID
    audio, err, detail = _tts_request(api_key, voice_id, text, model_id)
    if audio:
        return audio, ""
    if "library voices" in (detail or "").lower() and voice_id != FREE_PLAN_VOICE_ID:
        audio, err, detail = _tts_request(api_key, FREE_PLAN_VOICE_ID, text, model_id)
        if audio:
            return audio, ""
    return None, err or "tts_failed"


def transcribe_speech(audio_bytes: bytes, filename: str, content_type: str, language_code: str) -> tuple[str, str]:
    """ElevenLabs Scribe STT. Returns (transcript, error_message)."""
    api_key = getattr(settings, "ELEVENLABS_API_KEY", "")
    if not api_key:
        return "", "missing_key"
    try:
        data = {"model_id": "scribe_v2"}
        if language_code:
            data["language_code"] = language_code
        resp = requests.post(
            STT_URL,
            headers={"xi-api-key": api_key},
            files={"file": (filename or "speech.webm", audio_bytes, content_type or "audio/webm")},
            data=data,
            timeout=45,
        )
        resp.raise_for_status()
        payload = resp.json()
        return (payload.get("text") or "").strip(), ""
    except Exception as e:
        body = ""
        if "resp" in locals() and getattr(resp, "text", None):
            body = resp.text[:300]
        print(f"[ELEVENLABS STT FAILED] {e} {body}")
        return "", "stt_failed"

File: src/integrations.ts
import fs from "fs";
import path from "path";
import { parse } from "csv-parse/sync";

export interface WeatherData {
  mock: boolean;
  temp_c: number;
  condition: string;
  rain_chance_pct: number | null;
}

export async function getWeather(lat = 28.99, lon = 77.02): Promise<WeatherData> {
  const apiKey = process.env.OPENWEATHER_API_KEY;
  if (!apiKey) {
    return {
      mock: true,
      temp_c: 29,
      condition: "Partly cloudy",
      rain_chance_pct: 20,
    };
  }

  try {
    const url = `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}&units=metric`;
    const res = await fetch(url);
    if (!res.ok) throw new Error(`Weather API returned ${res.status}`);
    const data = (await res.json()) as any;
    return {
      mock: false,
      temp_c: Math.round(data.main?.temp ?? 29),
      condition: data.weather?.[0]?.description ?? "Clear",
      rain_chance_pct: null,
    };
  } catch (err) {
    console.warn("Weather fetch error, falling back to mock:", err);
    return {
      mock: true,
      temp_c: 29,
      condition: "Partly cloudy (fallback)",
      rain_chance_pct: 20,
    };
  }
}

export interface MarketPrice {
  crop: string;
  market: string;
  price_per_quintal: number;
  trend: string;
  trader_contact: string;
}

let cachedPrices: MarketPrice[] | null = null;

export function getMarketPrices(crop?: string): MarketPrice[] {
  if (!cachedPrices) {
    const csvPath = path.join(process.cwd(), "data", "market_prices.csv");
    if (!fs.existsSync(csvPath)) {
      // Fallback sample market prices if CSV not present
      cachedPrices = [
        { crop: "Wheat", market: "Sonipat Mandi", price_per_quintal: 2275, trend: "up", trader_contact: "+919811122233" },
        { crop: "Onion", market: "Nashik Mandi", price_per_quintal: 3100, trend: "down", trader_contact: "+919822233344" },
        { crop: "Rice", market: "Karnal Mandi", price_per_quintal: 3850, trend: "steady", trader_contact: "+919833344455" },
        { crop: "Cotton", market: "Rajkot Mandi", price_per_quintal: 7100, trend: "up", trader_contact: "+919844455566" },
      ];
    } else {
      try {
        const fileContent = fs.readFileSync(csvPath, "utf-8");
        const records = parse(fileContent, {
          columns: (header: string[]) => header.map((h) => h.trim().toLowerCase()),
          skip_empty_lines: true,
          bom: true,
        }) as Record<string, string>[];

        const list: MarketPrice[] = [];
        const trends = ["up", "down", "steady"];

        for (let i = 0; i < records.length; i++) {
          const row = records[i];
          const cropName = row.commodity || row.crop || "Unknown";
          const marketName = row.market || row.mandi || "Unknown";
          const priceRaw = row.modal_x0020_price || row.modal_price || row.price_per_quintal || row.price || "0";
          const price = parseFloat(priceRaw) || 0;
          const trend = row.trend || trends[i % trends.length];

          list.push({
            crop: cropName,
            market: marketName,
            price_per_quintal: price,
            trend: trend,
            trader_contact: row.trader_contact || "N/A",
          });
        }
        cachedPrices = list;
      } catch (e) {
        console.error("Error reading market_prices.csv:", e);
        cachedPrices = [];
      }
    }
  }

  if (crop) {
    const c = crop.toLowerCase();
    return cachedPrices.filter((p) => p.crop.toLowerCase() === c);
  }

  // When returning all prices for dashboard display:
  // Select key representative crops (Wheat, Onion, Rice, Cotton, Potato, Tomato)
  // or first 5 if none matched
  const keyCrops = ["wheat", "onion", "rice", "cotton", "potato", "tomato"];
  const selected: MarketPrice[] = [];
  const seenCrops = new Set<string>();

  for (const p of cachedPrices) {
    const cLow = p.crop.toLowerCase();
    if (keyCrops.includes(cLow) && !seenCrops.has(cLow)) {
      selected.push(p);
      seenCrops.add(cLow);
    }
  }

  if (selected.length > 0) {
    return selected;
  }
  return cachedPrices.slice(0, 5);
}

export async function sendAlert(phone: string, message: string): Promise<boolean> {
  const sid = process.env.TWILIO_ACCOUNT_SID;
  const token = process.env.TWILIO_AUTH_TOKEN;
  const from = process.env.TWILIO_WHATSAPP_FROM;

  if (!sid || !token) {
    console.log(`[MOCK ALERT] to ${phone}: ${message}`);
    return false;
  }

  try {
    const auth = Buffer.from(`${sid}:${token}`).toString("base64");
    const params = new URLSearchParams();
    params.append("To", `whatsapp:${phone}`);
    params.append("From", from || "");
    params.append("Body", message);

    const res = await fetch(`https://api.twilio.com/2010-04-01/Accounts/${sid}/Messages.json`, {
      method: "POST",
      headers: {
        Authorization: `Basic ${auth}`,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: params.toString(),
    });
    return res.ok;
  } catch (err) {
    console.error("[ALERT SEND FAILED]", err);
    return false;
  }
}

const TTS_URL = "https://api.elevenlabs.io/v1/text-to-speech";
const STT_URL = "https://api.elevenlabs.io/v1/speech-to-text";
const FREE_PLAN_VOICE_ID = "JBFqnCBsd6RMkjVDRZzb";

export async function synthesizeSpeech(text: string): Promise<{ audio: Buffer | null; error: string }> {
  const apiKey = (process.env.ELEVENLABS_API_KEY || "").trim().replace(/['"]/g, "");
  if (!apiKey) {
    return { audio: null, error: "missing_key" };
  }

  const voiceId = process.env.ELEVENLABS_VOICE_ID || FREE_PLAN_VOICE_ID;
  const modelId = process.env.ELEVENLABS_MODEL_ID || "eleven_multilingual_v2";

  try {
    const res = await fetch(`${TTS_URL}/${voiceId}`, {
      method: "POST",
      headers: {
        "xi-api-key": apiKey,
        "Content-Type": "application/json",
        Accept: "audio/mpeg",
      },
      body: JSON.stringify({
        text: text.slice(0, 2500),
        model_id: modelId,
      }),
    });

    if (res.ok) {
      const arrayBuf = await res.arrayBuffer();
      return { audio: Buffer.from(arrayBuf), error: "" };
    }

    const errText = await res.text();
    console.error("[ELEVENLABS TTS FAILED]", res.status, errText);
    return { audio: null, error: "tts_failed" };
  } catch (err: any) {
    console.error("[ELEVENLABS TTS FAILED]", err);
    return { audio: null, error: "tts_failed" };
  }
}

export async function transcribeSpeech(
  audioBuffer: Buffer,
  filename: string,
  contentType: string,
  languageCode = ""
): Promise<{ text: string; error: string }> {
  const apiKey = (process.env.ELEVENLABS_API_KEY || "").trim().replace(/['"]/g, "");
  if (!apiKey) {
    return { text: "", error: "missing_key" };
  }

  try {
    const formData = new FormData();
    const uint8 = new Uint8Array(audioBuffer);
    const blob = new Blob([uint8], { type: contentType || "audio/webm" });
    formData.append("file", blob, filename || "speech.webm");
    formData.append("model_id", "scribe_v2");
    if (languageCode) {
      formData.append("language_code", languageCode);
    }

    const res = await fetch(STT_URL, {
      method: "POST",
      headers: {
        "xi-api-key": apiKey,
      },
      body: formData,
    });

    if (res.ok) {
      const data = (await res.json()) as any;
      return { text: (data.text || "").trim(), error: "" };
    }

    const errBody = await res.text();
    console.error("[ELEVENLABS STT FAILED]", res.status, errBody);
    return { text: "", error: "stt_failed" };
  } catch (err: any) {
    console.error("[ELEVENLABS STT FAILED]", err);
    return { text: "", error: "stt_failed" };
  }
}
